File: api/routes/v1/operation_inbox.py
"""Cross-server inbox of queued, running, and retained operations."""

from __future__ import annotations

import asyncio
import json
from uuid import UUID

from fastapi import APIRouter, HTTPException, Request, status
from fastapi.responses import StreamingResponse

from api.dependencies import ActiveUser, DatabaseSession, StreamUser, close_request_session
from services.operations.inbox import (
    build_operation_inbox,
    list_accessible_servers,
    load_accessible_servers,
)
from services.operations.inbox_sse import InboxSseState, inbox_sse_headers, next_inbox_sse_frame
from services.operations.inbox_types import InboxItemData, InboxPayload
from services.plugins.ai_import_store import check_administrator, clear_failed_jobs
from services.plugins.description_sync_store import (
    clear_failed_jobs as clear_failed_description_jobs,
)
from services.server_operation_hub import server_operation_hub

from .operations import to_view
from .plugin_ai_imports import to_view as import_view
from .plugins import _description_sync_view as description_sync_view
from .schemas import ActionResult, OperationInboxItem, OperationInboxView

router = APIRouter(prefix="/api/v1/operations", tags=["v1-operations"])


def _to_inbox_item(item: InboxItemData) -> OperationInboxItem:
    view = to_view(item.record)
    return OperationInboxItem(
        **view.model_dump(),
        server_name=item.server_name,
        latest_message=item.latest_message,
        queue_position=item.queue_position,
    )


def to_inbox_view(payload: InboxPayload) -> OperationInboxView:
    items = [_to_inbox_item(item) for item in payload.items]
    completed_items = [_to_inbox_item(item) for item in payload.completed_items]
    failed_items = [_to_inbox_item(item) for item in payload.failed_items]
    running = [item for item in items if item.status == "running"]
    return OperationInboxView(
        market_import_items=[import_view(job) for job in payload.import_jobs],
        market_description_items=[description_sync_view(job) for job in payload.description_jobs],
        items=items,
        completed_items=completed_items,
        failed_items=failed_items,
        active_count=len(items),
        running_count=len(running),
        completed_count=len(completed_items),
        failed_count=len(failed_items),
        completed_retention_days=payload.completed_retention_days,
        failed_retention_days=payload.failed_retention_days,
    )


def encode_inbox_view(view: OperationInboxView) -> str:
    OperationInboxView.model_validate(view.model_dump())
    return json.dumps(
        view.model_dump(mode="json"),
        ensure_ascii=False,
        separators=(",", ":"),
        default=str,
    )


def _encode_inbox_payload(payload: InboxPayload) -> str:
    return encode_inbox_view(to_inbox_view(payload))


async def _build_inbox(
    servers: list[tuple[int, str]], include_imports: bool = False
) -> OperationInboxView:
    return to_inbox_view(await build_operation_inbox(servers, include_imports=include_imports))


@router.get("/inbox", response_model=OperationInboxView)
async def list_operation_inbox(
    db: DatabaseSession,
    current_user: ActiveUser,
) -> OperationInboxView:
    """Active jobs plus completed and failed history retained for seven days."""
    servers = await load_accessible_servers(db, current_user)
    await close_request_session(db)
    return await _build_inbox(servers, current_user.is_admin)


@router.get("/inbox/events", response_model=None)
async def stream_operation_inbox(
    request: Request,
    current_user: StreamUser,
) -> StreamingResponse:
    """Live inbox snapshots for the top-right tray. SSE, not a second WebSocket."""
    servers = await list_accessible_servers(current_user)

    async def event_source():
        yield ": connected\n\n"
        state = InboxSseState()
        while not await request.is_disconnected():
            include_imports = current_user.is_admin
            if include_imports:
                try:
                    await check_administrator(current_user.id)
                except PermissionError:
                    include_imports = False
            payload = await build_operation_inbox(servers, include_imports=include_imports)
            frame = next_inbox_sse_frame(state, payload, _encode_inbox_payload)
            if frame is not None:
                yield frame.text
            await asyncio.sleep(1)

    return StreamingResponse(
        event_source(),
        media_type="text/event-stream",
        headers=inbox_sse_headers(),
    )


@router.delete("/inbox/failed", response_model=ActionResult)
async def clear_failed_operations(
    db: DatabaseSession,
    current_user: ActiveUser,
) -> ActionResult:
    """Remove every retained failure the caller can see.

    Administrators also see failed AI marketplace imports in the same tray, so
    the one-click clear covers those too — they were otherwise undismissable
    and kept the tray badge red for their full seven-day retention.
    """
    servers = await load_accessible_servers(db, current_user)
    await close_request_session(db)
    cleared = await server_operation_hub.clear_failed([server_id for server_id, _name in servers])
    if current_user.is_admin:
        try:
            cleared += await clear_failed_jobs(current_user.id)
        except PermissionError:
            pass
        try:
            cleared += await clear_failed_description_jobs(current_user.id)
        except PermissionError:
            pass
    return ActionResult(
        success=True,
        message=f"Cleared {cleared} failed operation(s)",
    )


@router.delete("/inbox/failed/{operation_id}", response_model=ActionResult)
async def dismiss_failed_operation(
    operation_id: UUID,
    db: DatabaseSession,
    current_user: ActiveUser,
) -> ActionResult:
    """Remove one failed job from the retained failure tab."""
    allowed = {server_id for server_id, _name in await load_accessible_servers(db, current_user)}
    await close_request_session(db)
    record = await server_operation_hub.get(str(operation_id))
    if record is None or record.get("status") != "failed":
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Failed operation not found"
        )
    if int(record["server_id"]) not in allowed:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Failed operation not found"
        )
    dismissed = await server_operation_hub.dismiss_failed(str(operation_id))
    if dismissed is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Failed operation not found"
        )
    return ActionResult(success=True, message="Failed operation cleared")


@router.delete("/inbox/completed", response_model=ActionResult)
async def clear_completed_operations(
    db: DatabaseSession,
    current_user: ActiveUser,
) -> ActionResult:
    """Remove every retained successful job the caller can see."""
    servers = await load_accessible_servers(db, current_user)
    await close_request_session(db)
    cleared = await server_operation_hub.clear_completed(
        [server_id for server_id, _name in servers]
    )
    return ActionResult(
        success=True,
        message=f"Cleared {cleared} completed operation(s)",
    )


@router.delete("/inbox/completed/{operation_id}", response_model=ActionResult)
async def dismiss_completed_operation(
    operation_id: UUID,
    db: DatabaseSession,
    current_user: ActiveUser,
) -> ActionResult:
    """Remove one completed job from the retained history."""
    allowed = {server_id for server_id, _name in await load_accessible_servers(db, current_user)}
    await close_request_session(db)
    record = await server_operation_hub.get(str(operation_id))
    if record is None or record.get("status") != "completed":
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Completed operation not found"
        )
    if int(record["server_id"]) not in allowed:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Completed operation not found"
        )
    dismissed = await server_operation_hub.dismiss_completed(str(operation_id))
    if dismissed is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Completed operation not found"
        )
    return ActionResult(success=True, message="Completed operation cleared")

File: frontend/src/modules/servers/operation-inbox.ts
import {
  SERVER_OPERATION_ACTIONS,
  type OperationInbox,
  type OperationInboxItem,
  type ServerOperation,
  type ServerOperationAction,
  type ServerStatus,
} from "@/modules/servers/types";

export const OPERATION_INBOX_EVENTS_URL = "/ops-stream/operations/inbox";

const KNOWN_STATUSES: readonly ServerStatus[] = [
  "pending",
  "deploying",
  "running",
  "stopped",
  "error",
  "unknown",
];

type InboxItemDto = {
  operation_id: string;
  server_id: number;
  action: string;
  status: OperationInboxItem["status"];
  success?: boolean | null;
  message?: string | null;
  server_status?: string | null;
  started_at: string;
  completed_at?: string | null;
  actor_user_id: number;
  stream_url: string;
  command?: string | null;
  server_name: string;
  latest_message?: string | null;
  queue_position?: number;
};

export type InboxSnapshotDto = {
  market_import_items?: OperationInbox["marketImportItems"];
  items: InboxItemDto[];
  completed_items?: InboxItemDto[];
  failed_items?: InboxItemDto[];
  active_count: number;
  running_count: number;
  completed_count?: number;
  failed_count?: number;
  completed_retention_days?: number;
  failed_retention_days?: number;
};

export function mapServerStatus(value: string): ServerStatus {
  return (KNOWN_STATUSES as readonly string[]).includes(value)
    ? (value as ServerStatus)
    : "unknown";
}

function toOperationAction(value: string): ServerOperationAction {
  return (SERVER_OPERATION_ACTIONS as readonly string[]).includes(value)
    ? (value as ServerOperationAction)
    : "status";
}

type OperationViewDto = {
  operation_id: string;
  server_id: number;
  action: string;
  status: ServerOperation["status"];
  success?: boolean | null;
  message?: string | null;
  server_status?: string | null;
  started_at: string;
  completed_at?: string | null;
  actor_user_id: number;
  stream_url: string;
  command?: string | null;
};

export function mapServerOperation(raw: OperationViewDto): ServerOperation {
  return {
    operationId: raw.operation_id,
    serverId: raw.server_id,
    action: toOperationAction(raw.action),
    status: raw.status,
    success: raw.success ?? null,
    message: raw.message ?? null,
    serverStatus: raw.server_status ? mapServerStatus(raw.server_status) : null,
    startedAt: raw.started_at,
    completedAt: raw.completed_at ?? null,
    actorUserId: raw.actor_user_id,
    streamUrl: raw.stream_url,
    command: typeof raw.command === "string" ? raw.command : null,
  };
}

function toInboxItem(item: InboxItemDto): OperationInboxItem {
  return {
    ...mapServerOperation(item),
    serverName: item.server_name,
    latestMessage: item.latest_message ?? null,
    queuePosition: item.queue_position ?? 0,
  };
}

export function mapOperationInbox(raw: InboxSnapshotDto): OperationInbox {
  const completedItems = (raw.completed_items ?? []).map(toInboxItem);
  const failedItems = (raw.failed_items ?? []).map(toInboxItem);
  return {
    items: raw.items.map(toInboxItem),
    completedItems,
    marketImportItems: raw.market_import_items ?? [],
    failedItems,
    activeCount: raw.active_count,
    runningCount: raw.running_count,
    completedCount: raw.completed_count ?? completedItems.length,
    failedCount: raw.failed_count ?? failedItems.length,
    completedRetentionDays: raw.completed_retention_days ?? 7,
    failedRetentionDays: raw.failed_retention_days ?? 7,
  };
}

export function parseOperationInboxPayload(raw: string): OperationInbox | null {
  try {
    const data = JSON.parse(raw) as Partial<InboxSnapshotDto>;
    if (!Array.isArray(data.items) || typeof data.active_count !== "number") {
      return null;
    }
    return mapOperationInbox({
      items: data.items,
      completed_items: data.completed_items,
      market_import_items: data.market_import_items,
      failed_items: data.failed_items,
      active_count: data.active_count,
      running_count: data.running_count ?? 0,
      completed_count: data.completed_count,
      failed_count: data.failed_count,
      completed_retention_days: data.completed_retention_days,
      failed_retention_days: data.failed_retention_days,
    });
  } catch {
    return null;
  }
}
